File: CodeExchange/cx.lua
-- CodeExchange (cx) v3.0.0 — Lua Implementation

local function emit_raw_buffer(output, buffer)
    if #buffer == 0 then return end
    local crc = 0
    for i = 1, #buffer do
        crc = bit32.bxor(crc, string.byte(buffer, i))
    end
    output:write(buffer .. "|" .. string.format("%02X", crc) .. "\n")
end

local function flush_run(buffer, char, count, output, line_length)
    if count == 0 then return buffer end
    
    local escape_map = { ["\n"]="\\n", ["\t"]="\\t", ["\r"]="\\r", ["\\"]="\\\\" }
    local normal_seq = escape_map[char] or char
    local normal_width = #normal_seq
    local max_allowed = line_length > 4 and (line_length - 4) or line_length

    while count > 0 do
        local chunk = math.min(count, 65535)
        local raw_cost = normal_width * chunk
        local rle_cost = chunk > 255 and (6 + normal_width) or (4 + normal_width)

        if line_length > 0 and (#buffer + rle_cost > max_allowed) then
            if #buffer > 0 then
                emit_raw_buffer(output, buffer)
                buffer = ""
            end
        end

        if rle_cost < raw_cost then
            local marker = chunk > 255 and string.format("/z%04X", chunk) or string.format("/x%02X", chunk)
            buffer = buffer .. marker .. normal_seq
            count = count - chunk
        else
            local process_now = chunk
            if line_length > 0 then
                local space_left = math.max(0, max_allowed - #buffer)
                local max_raw = math.floor(space_left / normal_width)
                if max_raw == 0 then
                    emit_raw_buffer(output, buffer)
                    buffer = ""
                else
                    process_now = math.min(process_now, max_raw)
                end
            end
            buffer = buffer .. string.rep(normal_seq, process_now)
            count = count - process_now
        end

        if line_length > 0 and #buffer >= max_allowed then
            emit_raw_buffer(output, buffer)
            buffer = ""
        end
    end
    return buffer
end

function encode_file(output, file_path, line_length)
    local file = io.open(file_path, "rb")
    if not file then return end

    output:write("===== FILE: " .. file_path .. " - LL:" .. line_length .. " ====\n")
    local global_crc = 0
    local line_buffer = ""
    local current_ch = nil
    local run_count = 0

    while true do
        local chunk = file:read(16384)
        if not chunk then break end
        for i = 1, #chunk do
            local ch = string.sub(chunk, i, i)
            global_crc = bit32.bxor(global_crc, string.byte(ch))

            if run_count == 0 then
                current_ch = ch
                run_count = 1
            elseif ch == current_ch then
                run_count = run_count + 1
            else
                line_buffer = flush_run(line_buffer, current_ch, run_count, output, line_length)
                current_ch = ch
                run_count = 1
            end
        end
    end

    if run_count > 0 then line_buffer = flush_run(line_buffer, current_ch, run_count, output, line_length) end
    if #line_buffer > 0 then emit_raw_buffer(output, line_buffer) end

    output:write(string.format("===== ENDFILE: %s - FCS:%02X ====\n", file_path, global_crc))
    file:close()
end

File: CodeExchange/cx.cs
using System;
using System.IO;
using System.Text;

namespace CodeExchange
{
    /// <summary>
    /// CodeExchange (cx) v3.0.0 — C# Core Implementation
    /// Optimized stack-allocation structure utilizing modern file access windows.
    /// </summary>
    class Program
    {
        private const int IoBufferSize = 16384;
        private static readonly char[] HexDigits = "0123456789ABCDEF".ToCharArray();

        private static void EmitRawBuffer(StreamWriter output, byte[] buffer, int len)
        {
            if (len == 0) return;
            byte crc = 0;
            for (int i = 0; i < len; i++)
            {
                crc ^= buffer[i];
            }
            
            string payload = Encoding.ASCII.GetString(buffer, 0, len);
            output.Write(payload);
            output.Write('|');
            output.Write(HexDigits[(crc >> 4) & 0x0F]);
            output.Write(HexDigits[crc & 0x0F]);
            output.Write('\n');
        }

        private static int FlushRun(byte[] buffer, int bufLen, byte charByte, int count, StreamWriter output, int lineLength)
        {
            if (count == 0) return bufLen;

            byte[] normalSeq = charByte switch
            {
                10 => Encoding.ASCII.GetBytes("\\n"),
                9  => Encoding.ASCII.GetBytes("\\t"),
                13 => Encoding.ASCII.GetBytes("\\r"),
                92 => Encoding.ASCII.GetBytes("\\\\"),
                _  => new byte[] { charByte }
            };

            int normalWidth = normalSeq.Length;
            int maxAllowed = lineLength > 4 ? lineLength - 4 : lineLength;

            while (count > 0)
            {
                int chunk = Math.Min(count, 65535);
                int rawCost = normalWidth * chunk;
                int rleCost = chunk > 255 ? (6 + normalWidth) : (4 + normalWidth);

                if (lineLength > 0 && (bufLen + rleCost > maxAllowed))
                {
                    if (bufLen > 0)
                    {
                        EmitRawBuffer(output, buffer, bufLen);
                        bufLen = 0;
                        continue;
                    }
                }

                if (rleCost < rawCost)
                {
                    string marker = chunk > 255 ? $"/z{chunk:X4}" : $"/x{chunk:X2}";
                    byte[] markerBytes = Encoding.ASCII.GetBytes(marker);
                    
                    Array.Copy(markerBytes, 0, buffer, bufLen, markerBytes.Length);
                    bufLen += markerBytes.Length;
                    Array.Copy(normalSeq, 0, buffer, bufLen, normalSeq.Length);
                    bufLen += normalSeq.Length;
                    
                    count -= chunk;
                }
                else
                {
                    int processNow = chunk;
                    if (lineLength > 0)
                    {
                        int spaceLeft = Math.Max(0, maxAllowed - bufLen);
                        int maxRaw = spaceLeft / normalWidth;
                        if (maxRaw == 0)
                        {
                            EmitRawBuffer(output, buffer, bufLen);
                            bufLen = 0;
                            continue;
                        }
                        processNow = Math.Min(processNow, maxRaw);
                    }

                    for (int i = 0; i < processNow; i++)
                    {
                        Array.Copy(normalSeq, 0, buffer, bufLen, normalSeq.Length);
                        bufLen += normalSeq.Length;
                    }
                    count -= processNow;
                }

                if (lineLength > 0 && bufLen >= maxAllowed)
                {
                    EmitRawBuffer(output, buffer, bufLen);
                    bufLen = 0;
                }
            }
            return bufLen;
        }

        private static void EncodeFile(StreamWriter output, string filePath, int lineLength)
        {
            if (!File.Exists(filePath)) return;

            output.Write($"===== FILE: {filePath} - LL:{lineLength} ====\n");

            byte globalCrc = 0;
            byte[] lineBuffer = new byte[4096];
            int bufLen = 0;
            
            int runCount = 0;
            byte currentByte = 0;
            byte[] ioBuf = new byte[IoBufferSize];

            using (FileStream fs = File.OpenRead(filePath))
            {
                int bytesRead;
                while ((bytesRead = fs.Read(ioBuf, 0, ioBuf.Length)) > 0)
                {
                    for (int i = 0; i < bytesRead; i++)
                    {
                        byte b = ioBuf[i];
                        globalCrc ^= b;

                        if (runCount == 0)
                        {
                            currentByte = b;
                            runCount = 1;
                        }
                        else if (b == currentByte)
                        {
                            runCount++;
                        }
                        else
                        {
                            bufLen = FlushRun(lineBuffer, bufLen, currentByte, runCount, output, lineLength);
                            currentByte = b;
                            runCount = 1;
                        }
                    }
                }
            }

            if (runCount > 0) bufLen = FlushRun(lineBuffer, bufLen, currentByte, runCount, output, lineLength);
            if (bufLen > 0) EmitRawBuffer(output, lineBuffer, bufLen);

            output.Write($"===== ENDFILE: {filePath} - FCS:{globalCrc:X2} ====\n");
        }

        static void Main(string[] args)
        {
            if (args.Length < 2) return;

            if (args[0] == "-e")
            {
                int lineLimit = 80;
                int fileIdx = 1;
                if (args.Length > 2 && args[1] == "-l")
                {
                    int.TryParse(args[2], out lineLimit);
                    fileIdx = 3;
                }

                using (StreamWriter stdout = new StreamWriter(Console.OpenStandardOutput(), Encoding.ASCII))
                {
                    stdout.AutoFlush = true;
                    stdout.Write("CodeExchange format - xc format v3 (Single-Pass Stream Architecture)\n\n");
                    for (int i = fileIdx; i < args.Length; i++)
                    {
                        EncodeFile(stdout, args[i], lineLimit);
                    }
                    stdout.Write("===== EOF\n");
                }
            }
        }
    }
}
